### PR TITLE
 Allow enabling egress gateway feature

### LINKED ISSUE #98: Allow enabling egress gateway feature
**How to categorize this issue?**
/area networking
/kind enhancement

**What would you like to be added**:

Cilium introduced the Egress Gateway feature in v1.10 (see the [docs](https://docs.cilium.io/en/stable/gettingstarted/egress-gateway/)). The egress gateway allows users to redirect egress pod traffic through specific, gateway nodes. Packets are masqueraded to the gateway node IP. To make use of this feature with gardener this must be set on the configmap for cilium:

```yaml
enable-ipv4-egress-gateway: true
enable-bpf-masquerade: true
kube-proxy-replacement: strict
```

**Why is this needed**:

For some of our services running in K8s it is necessary to have a single egress IP since the external service ties sessions to IP addresses.

### PR DESCRIPTION
**Pull request by jastBytes**
**How to categorize this PR?**
/area networking
/kind enhancement

**What this PR does / why we need it**:
Egress gateway feature is now supported in cilium. It allows users to redirect egress pod traffic through specific, gateway nodes.
To enable the feature kube-proxy must be disabled.

```
spec:
  kubernetes:
    kubeProxy:
      enabled: false
```
The feature can be enabled via:

```
spec:
  networking:
    providerConfig:
      apiVersion: cilium.networking.extensions.gardener.cloud/v1alpha1
      kind: NetworkConfig
      egressGateway:
         enabled: true
```

**Which issue(s) this PR fixes**:
Fixes #98

**Special notes for your reviewer**:

**Release note**:

```other operator
Egress gateway feature is now supported in cilium. It allows users to redirect egress pod traffic through specific, gateway nodes.
```

### CODE REVIEW

**Comment by gardener-robot-ci-3** [comment]
Thank you @jastBytes for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

**Comment by christianhuening** [comment]
@ScheererJ would be awesome if you could have a look whether this is feasible. We would need that feature exposed for configuration. 
Thanks a ton!

**Review by DockToFuture — CHANGES_REQUESTED** [review]
Please also add in `pkg/apis/cilium/types_network.go` 	`EgressGateway *bool` and run `make generate` afterwards. 
Can you please also add some comment/docu in https://github.com/gardener/gardener-extension-networking-cilium/blob/master/docs/usage-as-end-user.md.

File: pkg/apis/cilium/types_network.go [hunk 1]
```diff
@@ -138,4 +138,7 @@ type NetworkConfig struct {
 	IPv6 *IPv6
 	// BPFSocketLBHostnsOnly flag to be enabled or not
 	BPFSocketLBHostnsOnly *BPFSocketLBHostnsOnly
+	// EgressGateway flag to be enabled or not
+	// +optional
```
Comment (DockToFuture) [inline pkg/apis/cilium/types_network.go:142]: this line can be removed

**Comment by DockToFuture** [comment]
Can you plead run `make generate` and add the generated files to the PR?

File: pkg/apis/cilium/types_network.go [hunk 2]
```diff
@@ -138,4 +138,6 @@ type NetworkConfig struct {
 	IPv6 *IPv6
 	// BPFSocketLBHostnsOnly flag to be enabled or not
 	BPFSocketLBHostnsOnly *BPFSocketLBHostnsOnly
+	// EgressGateway flag to be enabled or not
+	EgressGateway *bool
```
Comment (DockToFuture) [inline pkg/apis/cilium/types_network.go:142]:
```suggestion
	// EgressGateway flag to be enabled or not
	EgressGateway *EgressGateway
```

// EgressGateway enablement for cilium
type EgressGateway struct {
	Enabled bool
}

File: pkg/apis/cilium/v1alpha1/types_network.go [hunk 3]
```diff
@@ -138,4 +138,7 @@ type NetworkConfig struct {
 	// BPFSocketLBHostnsOnly flag to be enabled or not
 	// +optional
 	BPFSocketLBHostnsOnly *BPFSocketLBHostnsOnly `json:"bpfSocketLBHostnsOnly,omitempty"`
+	// EgressGateway flag to be enabled or not
+	// +optional
+	EgressGateway *bool `json:"egressGateway,omitempty"`
```
Comment (DockToFuture) [inline pkg/apis/cilium/v1alpha1/types_network.go:143]:
```suggestion
	EgressGateway *EgressGateway `json:"egressGateway,omitempty"`
```

and add 
```
// EgressGateway enablement for cilium
type EgressGateway struct {
	Enabled bool `json:"enabled"`
}
```

File: pkg/charts/utils.go [hunk 4]
```diff
@@ -189,5 +192,10 @@ func generateChartValues(config *ciliumv1alpha1.NetworkConfig, network *extensio
 		globalConfig.Debug.Enabled = *config.Debug
 	}
 
+	// check if egress gateway is enabled
+	if config.EgressGateway != nil {
+		globalConfig.EgressGateway.Enabled = *config.EgressGateway
```
Comment (DockToFuture) [inline pkg/charts/utils.go:197]:
```suggestion
		globalConfig.EgressGateway.Enabled = config.EgressGateway.Enabled
```

**Comment by DockToFuture** [comment]
> Can you please run `make generate` and add the generated files to the PR?

Still needs to be done. :)

**Comment by jastBytes** [comment]
`make generate` ran through just fine, but no changes. Any hint?

```
> Generate
Successfully generated controller registration at ../../example/controller-registration.yaml
Generating deepcopy funcs
Generating defaulters
Generating conversions
Generating deepcopy funcs
Generating defaulters
Generating conversions
I0629 14:44:09.768812   35224 main.go:127] parsing go packages in directory .
I0629 14:44:13.903833   35224 main.go:229] using package=.
I0629 14:44:13.915682   35224 main.go:165] written to ../../../../hack/api-reference/cilium.md
I0629 14:44:13.933826   35297 main.go:127] parsing go packages in directory .
I0629 14:44:25.345737   35297 main.go:229] using package=.
I0629 14:44:25.763154   35297 main.go:165] written to ../../../../hack/api-reference/config.md
```

**Comment by DockToFuture** [comment]
There should be some changes in the generated files because of type changes made under `pkg/apis/cilium/*`. The printed output seems to be ok for the `make generate` cmd.

**Comment by DockToFuture** [comment]
Hello @jastBytes, is this PR still relevant to you?

**Comment by jastBytes** [comment]
Excuse me! I really forgot about this one. We found that we currently do not need this feature since the egress IP address is a single one for shoots at GCP. I still would like to give you access because maybe someone else profits and even maybe we run into the need for this in the future. As far as I understood for AWS there are different egress IP addresses for the same cluster per zone.

**Comment by DockToFuture** [comment]
For gcp it must not necessarily be one static IP! But if you want to have one or more static IP addresses you could use the CloudNAT feature and manually assign the IP addresses to the CloudNAT and specify the min port size used per VM. See https://github.com/gardener/gardener-extension-provider-gcp/blob/master/docs/usage-as-end-user.md. With that you can assure at least that the used IP addresses remain static otherwise they might change.

**Review by DockToFuture — APPROVED** [review]
/lgtm

**Pull request merged**